### PR TITLE
Add autoregressiveness to MHA (again)

### PR DESCRIPTION
**Pull request by Artur-Galstyan**
Hi, I noticed that the autoregressive-attention branch has become stale and a few things have changed in the dev branch since then. So I thought I'd revive the auto regressive branch.

Note that this does not yet have the MQA <-> MHA interpolation. In our last discussion, I told you about my "double VMAP" strategy and that LLMs such as LLaMA don't use that but instead simply "repeat" the keys and values until they match the queries. You were not sure which approach to follow.

I actually have a working version of the MHA implementation in which I allow the user to choose which one they want (see [this](https://github.com/Artur-Galstyan/kira/blob/main/kira/modules/mha.py#L136). Maybe in the future we could have something similar, but for now, I think it's ok to add autoregressive decoding - even without the MHA<->MQA interpolation. 

WDYT?

### CODE REVIEW

File: equinox/nn/_attention.py [hunk 1]
```diff
@@ -215,6 +237,60 @@ def __init__(
         self.use_key_bias = use_key_bias
         self.use_value_bias = use_value_bias
         self.use_output_bias = use_output_bias
+        self.autoregressive_index = StateIndex(_make_autoregressive_cache())
```
Comment (patrick-kidger) [inline equinox/nn/_attention.py:240]: If `state_length is None` then I think we'll hit the error message here -- so I think this will unconditionally raise?

File: equinox/nn/_attention.py [hunk 2]
```diff
@@ -323,10 +438,53 @@ def __call__(
...
+        else:
+            return out, state
 
     def _project(self, proj, x):
         seq_length, _ = x.shape
         projection = jax.vmap(proj)(x)
         return projection.reshape(seq_length, self.num_heads, -1)
+
+
+def self_attention(
```
Comment (patrick-kidger) [inline equinox/nn/_attention.py:454]: I think this function was most useful back when we had MQA, and for which the `MultiheadAttention` API was a bit too much.

I don't think we're *quite* there yet (although it's definitely still pretty complicated), so I think I'd probably suggest removing this function for now.

**Review by patrick-kidger — COMMENTED** [review]
Thank you for this! This would be awesome to get in.

Would you be interested in writing some tests? Frankly the implementation of this is fairly complicated. I think I mostly trust past-me when I wrote this... but that not *that* much!

File: equinox/nn/_attention.py [hunk 3]
```diff
@@ -308,6 +390,39 @@ def __call__(
...
+        if state is sentinel:
+            causal_mask_offset = 0
+        else:
+            key_state, value_state, index = state.get(self.autoregressive_index)
+            key_state = lax.dynamic_update_slice_in_dim(
+                key_state, key_heads, index, axis=0
+            )
+            value_state = lax.dynamic_update_slice_in_dim(
+                value_state, value_heads, index, axis=0
+            )
```
Comment (patrick-kidger) [inline equinox/nn/_attention.py:402]: By the way, I think there is actually a subtle bug in these two lines: when `index` is close to the end, then it gets pushed back so that all of the inserted array is inserted (rather than simply truncating the inserted array to only insert the bits that fit).

I don't know if that's something that can be resolved efficiently, though.

Comment (Artur-Galstyan) [reply equinox/nn/_attention.py:402]: Ah I see. Well, the solution is probably to truncate `key|value_state` to account for that before passing them into the `dynamic_update_slice_in_dim` function. I'll see what I can do here.

Comment (Artur-Galstyan) [reply equinox/nn/_attention.py:402]: Alternatively, we might want to employ the FiFo strategy and roll the key and value state arrays to make room for the new entries.

Comment (Artur-Galstyan) [reply equinox/nn/_attention.py]: I've added rolling; but I'll need to train something to see if it still works fine this way. I'll keep you posted.

File: equinox/nn/_attention.py [hunk 4]
```diff
@@ -308,6 +391,58 @@ def __call__(
...
+        else:
+            if self.autoregressive_index is None:
+                raise ValueError(
+                    "State was provided, but cannot use autoregressive decoding "
+                    "without specifying "
+                    "`MultiheadAttention(..., state_length=...)`."
+                )
+            key_state, value_state, index = state.get(self.autoregressive_index)
+            key_seq_length = key_heads.shape[0]
+            if index + key_seq_length > self.state_length:
```
Comment (patrick-kidger) [inline equinox/nn/_attention.py:405]: These `if` statements will fail under JIT.

FWIW I suspect it won't be possible to make things work efficiently at the end anyway, and it's not obvious to me that rolling is really the better choice. (E.g. some techniques make sure to use the start of an attention window as a place to put globally-attended-to information.) I think I'm probably happy with the behaviour without this.

Comment (Artur-Galstyan) [reply equinox/nn/_attention.py]: I see. I reverted it back to the previous implementation.

AFAICT in the current implementation of the cache, if the cache is:

`1 2 3 4 5 6 7 0 0`

and the `seq_len` is 3, then after the update it becomes:

`1 2 3 4 5 6 A B C` 

But if I understood you correctly (from your previous comment), you want it to be 

`1 2 3 4 5 6 7 A B` 

right? In the rolling strategy, it would have become:

`2 3 4 5 6 7 A B C`.

 Perhaps David could weigh in on this, because I don't know what the best approach is. @dlwh WDYT?

Comment (patrick-kidger) [reply equinox/nn/_attention.py]: A very neat explanation! Yes, that's it exactly.

Comment (dlwh) [reply equinox/nn/_attention.py]: I've stayed away from inference so far, so I mostly haven't had to think about caching. That seems sensible and clean.

It does seem like a thing this could be pluggable? I know it's built on State right now, but if the KVCache were a module with an abstract method, you could choose the behavior.

The MaxText code is pretty hard to follow (for me) but I think there are two caches there, one that does the prefilling/prefix that Patrick talks about, and another that's for the ar decoding. [this](https://github.com/google/maxtext/blob/1ae752103af6f2939fbb3f6bbbfa998eee8e4d76/MaxText/layers/attentions.py#L653) means it's doing a roll for AR tokens and leaving the prefill tokens in place. I think?

Comment (Artur-Galstyan) [reply equinox/nn/_attention.py]: > It does seem like a thing this could be pluggable? I know it's built on State right now, but if the KVCache were a module with an abstract method, you could choose the behavior.

I really like this idea! I can cook something up. It would also make the already pretty large MHA file slightly more svelte.

**Comment by Artur-Galstyan** [comment]
I've added a first draft for an `AbstractKVCache`. The idea is that people can define their own KVCache (if they wish to) and pass that to the MHA constructor. The abstract class calls the abstract method in `__call__` and does some checks around it to make sure that it was implemented correctly. From there, the MHA implementation just takes the new KV matrices from the KV cache and works with those. 

I also provided a "StandardKVCache" which is essentially what we had before (without truncation). I was thinking of adding another one which does rolling and then let the user choose which one they want -- or let them create their own.

LMK what you think of this approach.

**Comment by patrick-kidger** [comment]
I didn't close this intentionally -- deleting the target branch (in this case, `dev`), causes all PRs against it to be automatically closed. Feel free to reopen this against `main`!

File: equinox/nn/_attention.py [hunk 5]
```diff
@@ -144,12 +148,14 @@ def __init__(
...
         qk_size: Optional[int] = None,
         vo_size: Optional[int] = None,
+        state_length: Optional[int] = None,
         use_query_bias: bool = False,
         use_key_bias: bool = False,
         use_value_bias: bool = False,
         use_output_bias: bool = False,
         dropout_p: float = 0.0,
         inference: bool = False,
+        kv_cache: Optional[AbstractKVCache] = None,
```
Comment (patrick-kidger) [inline equinox/nn/_attention.py:158]: This argument doesn't seem to be documented in the docstring.

File: equinox/nn/_attention.py [hunk 6]
```diff
@@ -308,6 +380,39 @@ def __call__(
...
+                    "See `equinox.nn.StandardKVCache` for an example."
+                )
+
+            key_state, value_state, index, state = self.kv_cache(
+                state, key_heads, value_heads
+            )
+            causal_mask_offset = index - kv_seq_length
+            key_heads = key_state
+            value_heads = value_state
+            kv_seq_length = self.state_length
```
Comment (patrick-kidger) [inline equinox/nn/_attention.py:400]: I like this! It helps to make the `MultiheadAttention.__call__` code a little less of a gigantic monolith.

File: equinox/nn/_attention.py [hunk 7]
```diff
@@ -256,6 +325,9 @@ def __call__(
             be a JAX array of shape `(query_seq_length, kv_seq_length)`, or (for custom
```
Comment (patrick-kidger) [inline equinox/nn/_attention.py:325]: The `mask="causal"` option needs documenting.

I suspect this might end up being quite long. Maybe we should pull it out into an admonition (the `!!! foo` things you see in the docs). As a reference point, se how we handle the discussion on padding for `eqx.nn.Conv`.

File: equinox/nn/_attention.py (hunk 5 above)
Comment (patrick-kidger) [reply equinox/nn/_attention.py:158]: Also, I think it would be really useful to have a short example somewhere -- maybe at the bottom of the `__init__` docstring? Not a strong feeling on location -- on how to use `MultiheadAttention` with state and kv caching.

Comment (patrick-kidger) [reply equinox/nn/_attention.py:158]: Also I think `state_length` is unused here now? It appears solely in the `kv_cache`. We should remove it as an argument.

File: equinox/nn/_kv_cache.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+import jax.lax as lax
+import jax.numpy as jnp
+from jaxtyping import Array, Float, Int
+
+from .._misc import default_int_dtype
+from .._module import field, Module
+from ._stateful import State, StateIndex
+
+
+class AbstractKVCache(Module, strict=True):
```
Comment (patrick-kidger) [inline equinox/nn/_kv_cache.py:12]: So I suspect this doesn't need to actually be a full-blown abstract class. It looks like all the caller needs is a `__call__` with the appropriate API.

I think this means that we could have
```python
class MultiheadAttention(eqx.Module:
    def __init__(..., kv_cache: Optional[Callable[...signature here...]] = None):

class StandardKVCache(eqx.Module): ...
```
with the latter just collapsing the existing implementation of `AbstractKVCache` and `StandardKVCache` into just one class.

For bonus points this would mean we achieve consistency with the API for `process_heads`, which also just uses a `Callable` rather than an ABC. (In fact I don't think we really have ABCs anywhere else in `eqx.nn` -- neural networks seem to be simply enough that they're not needed.)

Comment (patrick-kidger) [reply equinox/nn/_kv_cache.py:12]: As a side note, some comment on the design as you currently have it (just for your own interest):
- `AbstractKVCache` is defining an `autoregressive_index` attribute, but this isn't actually used anywhere that `AbstractKVCache` is accepted. Rather, it is only used inside the `StandardKVCache` subclass. So this should be moved into the subclass.
- You might [this doc](https://docs.kidger.site/equinox/pattern/) interesting! It's the pattern I'm trying to use in Equinox and its ecosystem going forward. In particualr it advocates against `super`.

**Review by patrick-kidger — COMMENTED** [review]
Okay, I really like this as an idea! This had not occured to me at all, but I think it makes the code much simpler to read.

I've left some comments on how we might neated the structure some more, but let's definitely do this.
